api.php

```diff
@@ -0,0 +1,349 @@
+		return $response->withStatus(400);
+	}
+	// otherwise get the user
+	$user = User::findUserById($id);
+	// if we got nothing back from the user
+	if (empty($user)) {
+		// set not found message
+		$response->getBody()->write(json_encode([
+			'message' => 'user not found'
+		]));
+		// return 404
+		return $response
+			->withHeader('Content-Type', 'application/json')
+			->withStatus(404);
+	}
+	// assuming everything else went ok encode the user
+	$response->getBody()->write(json_encode($user->jsonSerialize()));
+	// return users info
+	return $response
+		->withHeader('Content-Type', 'application/json')
+		->withStatus(200);
+});
+
+// Create new user
+$app->post('/user/new', function (Request $request, Response $response, array $args) {
+	// get request body
+	$body = json_decode($request->getBody());
+	// create user from request values
+	$user = new User($body->first_name, $body->last_name, $body->email, $body->password, $body->phone, $body->preferred ?? false);
+	// save the user
+	$user->saveToDB();
+	// set encoded user as response body
+	$response->getBody()->write(json_encode($user));
+	// return response with 201 (added ok) code
+	return $response
+		->withHeader('Content-Type', 'application/json')
+		->withStatus(201);
+});
+
+//Change info for a user that already exists
+$app->post('/user/{id}', function (Request $request, Response $response, array $args) {
+	// get int value of requested id
+	$id = intval($args['id']);
+	// check validity of id
+	if (!$id) {
+		// set a message to explain what broke
+		$response->getBody()->write(json_encode([
+			'message' => 'invalid id provided',
+		]));
+		// return the error and a invalid request status
+		return $response->withStatus(400);
+	}
+	// look up the user
+	$user = User::findUserById($id);
+	// if no matching user found
+	if (!$user) {
+		// set the response message
+		$response->getBody()->write(json_encode([
+			'message' => 'no user found for id:' . $id,
+		]));
+		// return the error and a 404 status
+		return $response->withStatus(404);
+	}
+
+	// get the request as a stdObject
+	$body = json_decode($request->getBody());
+
+	//Update user info
+	if (isset($body->first_name)) {
+		$user->setFirstName($body->first_name);
+	}
+
+	if (isset($body->last_name)) {
+		$user->setLastName($body->last_name);
+	}
+
+	if (isset($body->email)) {
+		// TODO check if valid email
+		$user->setEmail($body->email);
+	}
+
+	if (isset($body->password)) {
+		$user->setPassword($body->password);
+	}
+
+	if (isset($body->phone)) {
+		$user->setPhoneNumber($body->phone);
+	}
+
+	// save the changes made to the db
+	$user->saveToDB();
+
+	// sent the response
+	$response->getBody()->write(json_encode($user));
+	// return response with correct code
+	return $response
+		->withHeader('Content-Type', 'application/json')
+		->withStatus(201);
+});
+
+//user/{userId}  delete a review by id
+$app->delete('/user/{userId}', function (Request $request, Response $response, array $args) {
```

Comment:
I'm getting a " User::$archivedAt must not be accessed before initialization " error message and it takes you to the jsonSerialize() function at the end of User.php

Reply:
this should be fixed